### ISSUE TITLE
Potential use for Docker HEALTHCHECK

### ISSUE THREAD

**Issue by ddavison** [SeleniumHQ/docker-selenium#1068]
## 🚀 Feature Proposal

We might be able to use Docker's [HEALTHCHECK](https://docs.docker.com/engine/reference/builder/#healthcheck) directive to check that the hub and/or nodes are up and running 🤔 

## Motivation

Instead of the recommended bash script, we might be able to use this.  Using a Health check might help with service providers like Kubernetes that use `--restart-always` directives.

## Example

```
HEALTHCHECK --interval=5m --timeout=3s
  CMD curl -f http://localhost/hub || exit 1
``` 

**Full disclosure - I've not used this directive before, so I'm not sure it will suit.  This directive was added after these images were created**

**Comment by diemol** [comment]
What do you think from this approach?
https://github.com/SeleniumHQ/docker-selenium#adding-a-healthcheck-to-the-grid

**Comment by ddavison** [comment]
The approach is certainly viable, but in the long-run, recommending a user of docker-selenium to add this healthcheck themselves might be a bit heavy.

Instead, the health of the container is determined by this HEALTHCHECK directive.

I envision something like this:

```
$ docker run -d --name hub selenium/hub
$ docker ps
selenium/hub STARTING
$ docker ps
selenium/hub STARTING
...

$ docker ps
selenium/hub RUNNING
```

Whereas that RUNNING is determined by the Healthcheck 👍

**Comment by diemol** [comment]
A PR would be cool :)

**Comment by brendon-r** [comment]
I'm pretty unfamiliar with the project so posting this here in case someone more familiar could give it a go. The healthcheck, *I think* should simply be able to call the check-grid script. Something like this in the base dockerfile:

```Dockerfile
FROM ubuntu:bionic-20200630
LABEL authors="Selenium <selenium-developers@googlegroups.com>"

#================================================
# Customize sources for apt-get
#================================================
RUN  echo "deb http://archive.ubuntu.com/ubuntu bionic main universe\n" > /etc/apt/sources.list \
  && echo "deb http://archive.ubuntu.com/ubuntu bionic-updates main universe\n" >> /etc/apt/sources.list \
  && echo "deb http://security.ubuntu.com/ubuntu bionic-security main universe\n" >> /etc/apt/sources.list

# No interactive frontend during docker build
ENV DEBIAN_FRONTEND=noninteractive \
    DEBCONF_NONINTERACTIVE_SEEN=true

#========================
# Miscellaneous packages
# Includes minimal runtime used for executing non GUI Java programs
#========================
RUN apt-get -qqy update \
  && apt-get -qqy --no-install-recommends install \
    bzip2 \
    ca-certificates \
    openjdk-8-jre-headless \
    tzdata \
    sudo \
    unzip \
    wget \
    jq \
    curl \
    supervisor \
    gnupg2 \
  && rm -rf /var/lib/apt/lists/* /var/cache/apt/* \
  && sed -i 's/securerandom\.source=file:\/dev\/random/securerandom\.source=file:\/dev\/urandom/' ./usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/java.security

#===================
# Timezone settings
# Possible alternative: https://github.com/docker/docker/issues/3359#issuecomment-32150214
#===================
ENV TZ "UTC"
RUN echo "${TZ}" > /etc/timezone \
  && dpkg-reconfigure --frontend noninteractive tzdata

#========================================
# Add normal user and group with passwordless sudo
#========================================
RUN groupadd seluser \
         --gid 1201 \
  && useradd seluser \
         --create-home \
         --gid 1201 \
         --shell /bin/bash \
         --uid 1200 \
  && usermod -a -G sudo seluser \
  && echo 'ALL ALL = (ALL) NOPASSWD: ALL' >> /etc/sudoers \
  && echo 'seluser:secret' | chpasswd
ENV HOME=/home/seluser

#======================================
# Add Grid check script
#======================================
COPY check-grid.sh entry_point.sh /opt/bin/

#======================================
# Add Supervisor configuration file
#======================================
COPY supervisord.conf /etc

#==========
# Selenium & relaxing permissions for OpenShift and other non-sudo environments
#==========
RUN  mkdir -p /opt/selenium /var/run/supervisor /var/log/supervisor \
  && touch /opt/selenium/config.json \
  && chmod -R 777 /opt/selenium /var/run/supervisor /var/log/supervisor /etc/passwd \
  && wget --no-verbose https://selenium-release.storage.googleapis.com/4.0-alpha-6/selenium-server-4.0.0-alpha-6.jar \
    -O /opt/selenium/selenium-server.jar \
  && chgrp -R 0 /opt/selenium ${HOME} /var/run/supervisor /var/log/supervisor \
  && chmod -R g=u /opt/selenium ${HOME} /var/run/supervisor /var/log/supervisor

#===================================================
# Run the following commands as non-privileged user
#===================================================
USER 1200:1201


CMD ["/opt/bin/entry_point.sh"]

HEALTHCHECK --interval=5s CMD /opt/bin/check-grid.sh

```

**Comment by diemol** [comment]
I was reviewing the HEALHCHECK docs we have in the README, and as mentioned in the previous comment, it works well. We also document how to create a HEALTHCHECK when starting a container.

I would not like to add HEALHCHECK instruction to the docker file because users could have different values for it. Some of them would like to check every 30 seconds, set retries, or something like that. I feel it is a good approach to point how to do it, and if a user thinks it is a good idea to add it, they can always create their own docker image.

In short, adding this option to the image itself would make it a bit less generic, which is the general idea of the project.

@ddavison I think this approach `CMD curl -f http://localhost/hub || exit 1` is not enough, some logic is needed to actually say that the grid is ready. Perhaps we are mixing HEALHCHECK with the concept of having the Grid ready?

If we want an endpoint where something like Kubernetes can check if the component is up, Grid 4 already includes an endpoint called `readyz`. If that what you are looking for?

**Comment by diemol** [comment]
I will close this because the issue went stale, if there is more motivation to dig deeper into this, feel free to comment and we can reopen and revisit the current approach.
